File: code/python/flyweight/Library.py
from ConcreteBook import ConcreteBook


class Library:
    def __init__(self):
        self.books = {}

    def get_book(self, title, author):
        if title not in self.books:
            new_id = len(self.books) + 1
            self.books[title] = ConcreteBook(title, author, new_id)
        return self.books[title]

    def borrow_book(self, title, author):
        book = self.get_book(title, author)
        if book.is_available():
            book.borrow_book()
        else:
            print(f"Book is currently not available: {title}. Ordering a new copy.")
            new_id = len(self.books) + 1
            new_book = ConcreteBook(title, author, new_id)
            self.books[title + " #" + str(new_id)] = new_book
            new_book.borrow_book()

    def return_book(self, title):
        if title in self.books:
            self.books[title].return_book()

File: code/typescript/src/flyweight/Library.ts
import { Book } from "./Book";
import { ConcreteBook } from "./ConcreteBook";

export class Library {
    private books: Record<string, Book> = {};

    getBook(title: string, author: string): Book {
        if (!this.books[title]) {
            this.books[title] = new ConcreteBook(title, author, Object.keys(this.books).length + 1);
        }
        return this.books[title];
    }

    borrowBook(title: string, author: string): void {
        const book = this.getBook(title, author);
        if (book.isAvailable()) {
            book.borrowBook();
        } else {
            const newBook = new ConcreteBook(title, author, Object.keys(this.books).length + 1);
            this.books[title] = newBook;
            console.log(`New copy ordered for: ${title} book id: ${book.getId() + 1}`);
            newBook.borrowBook();
        }
    }

    returnBook(title: string): void {
        const book = this.books[title];
        if (book) {
            book.returnBook();
        } else {
            console.log(`Book with title ${title} not found in the library.`);
        }
    }
}
